src/nat/probe.ts

```diff
@@ -0,0 +1,143 @@
+/// <reference path='../../../third_party/typings/es6-promise/es6-promise.d.ts' />
+/// <reference path='../../../third_party/freedom-typings/udp-socket.d.ts' />
+
+import arraybuffers = require('../arraybuffers/arraybuffers');
+import logging = require('../logging/logging');
+
+// NAT detection needs the help of an external server.
+// We keep this running on Amazon EC2 and the source
+// may be found here:
+//   https://github.com/uProxy/uproxy-probe/blob/master/python-src/probe-server.py
+var TEST_SERVER = '54.68.73.184';
+var TEST_PORT = 6666;
+
+var log :logging.Log = new logging.Log('probe');
+
+export function probe() : Promise<string> {
+  return new Promise((F, R) => {
+    var socket :freedom_UdpSocket.Socket = freedom['core.udpsocket']();
+    // The weird-looking type is due to a DefinitelyTyped weirdness
+    // mentioned here:
+    //   https://github.com/Microsoft/TypeScript/issues/842
+    var timerId :NodeJS.Timer = undefined;
+
+    var rejectShortcut: (e: any) => void = null;
+
+    function onUdpData(info: freedom_UdpSocket.RecvFromInfo) {
+      var rspStr: string = arraybuffers.arrayBufferToString(info.data);
+      log.debug('receive response = ' + rspStr);
+
+      var rsp = JSON.parse(rspStr);
+
+      if (rsp['answer'] == 'FullCone') {
```

Comment:
Yep, fixed. As discussed, I've filed an issue to perform a deeper cleanup:
https://github.com/uProxy/uproxy/issues/1808